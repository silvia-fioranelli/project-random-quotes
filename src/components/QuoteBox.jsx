import { shape, func, bool, string } from 'prop-types';

function QuoteBox(props) {
  const { quote, getQuote, isError } = props;

  const renderLoading = () => (
    <div class="spinner-border text-dark" role="status">
      <span class="sr-only">Loading...</span>
    </div>
  );

  const renderError = () => (
    <>
      <blockquote class="blockquote">
        <p class="mb-0">Ops! Something went wrong...</p>
      </blockquote>
      <hr />
      <button onClick={getQuote} className="btn btn-light">
        <i className="fa fa-refresh"></i>
      </button>
    </>
  );

  const renderQuote = () => (
    <>
      <blockquote class="blockquote">
        <p class="mb-0">{quote.content}</p>
        <br />
        <footer class="blockquote-footer font-italic">{quote.author}</footer>
      </blockquote>
      <hr />
      <button onClick={getQuote} className="btn btn-light">
        <i className="fa fa-refresh"></i>
      </button>
    </>
  );

  const render = () => {
    if (isError) {
      return renderError();
    }

    if (quote) {
      return renderQuote();
    }

    return renderLoading();
  };

  return (
    <div class="card text-center" style={{ width: '30rem' }}>
      <div class="card-body">{render()}</div>
    </div>
  );
}

QuoteBox.propTypes = {
  quote: shape({
    author: string.isRequired,
    content: string.isRequired,
  }),
  isError: bool.isRequired,
  getQuote: func.isRequired,
};

export default QuoteBox;
