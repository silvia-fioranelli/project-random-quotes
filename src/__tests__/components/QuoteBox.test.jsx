import { shallow } from 'enzyme';
import QuoteBox from '../../components/QuoteBox';

describe('QuoteBox', () => {
  const defaultProps = {
    getQuote: jest.fn(),
    isError: false,
    quote: {
      author: 'mock author',
      content: 'mock content',
    },
  };
  const render = (props) => shallow(<QuoteBox {...defaultProps} {...props} />);

  describe('rendering', () => {
    test('should render quote', () => {
      const wrapper = render();
      expect(wrapper).toMatchSnapshot();
    });

    test('should render loading', () => {
      const props = {
        quote: null,
      };
      const wrapper = render(props);
      expect(wrapper).toMatchSnapshot();
    });

    test('should render error', () => {
      const props = {
        isError: true,
      };
      const wrapper = render(props);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    test('refresh button should call getQuote', () => {
      const props = {
        getQuote: jest.fn(),
      };
      const wrapper = render(props);
      wrapper.find('button').simulate('click');
      expect(props.getQuote).toHaveBeenCalledTimes(1);
    });
  });
});
