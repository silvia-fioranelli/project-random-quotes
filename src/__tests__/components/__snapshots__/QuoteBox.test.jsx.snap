// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuoteBox rendering should render error 1`] = `
<div
  class="card text-center"
  style={
    Object {
      "width": "30rem",
    }
  }
>
  <div
    class="card-body"
  >
    <blockquote
      class="blockquote"
    >
      <p
        class="mb-0"
      >
        Ops! Something went wrong...
      </p>
    </blockquote>
    <hr />
    <button
      className="btn btn-light"
      onClick={[MockFunction]}
    >
      <i
        className="fa fa-refresh"
      />
    </button>
  </div>
</div>
`;

exports[`QuoteBox rendering should render loading 1`] = `
<div
  class="card text-center"
  style={
    Object {
      "width": "30rem",
    }
  }
>
  <div
    class="card-body"
  >
    <div
      class="spinner-border text-dark"
      role="status"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`QuoteBox rendering should render quote 1`] = `
<div
  class="card text-center"
  style={
    Object {
      "width": "30rem",
    }
  }
>
  <div
    class="card-body"
  >
    <blockquote
      class="blockquote"
    >
      <p
        class="mb-0"
      >
        mock content
      </p>
      <br />
      <footer
        class="blockquote-footer font-italic"
      >
        mock author
      </footer>
    </blockquote>
    <hr />
    <button
      className="btn btn-light"
      onClick={[MockFunction]}
    >
      <i
        className="fa fa-refresh"
      />
    </button>
  </div>
</div>
`;
